File: paleo_workbench/prediction/model_package.py
"""Production model package contract: validate + register (no training).

A package is a JSON/YAML-free plain dict or JSON file with:

- model_id, model_version, model_name, capability, provider
- artifact path + optional checksum
- input_schema / output_schema
- runtime, preprocessing_version, deterministic
- spatial_output_type
- scientific / demo_only flags

Registration uses the existing catalog Model/ModelVersion APIs.
Demo and heuristic providers cannot be registered as production packages.
"""

from __future__ import annotations

import hashlib
import json
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any

from paleo_workbench.catalog.models import Model, ModelVersion
from paleo_workbench.prediction.providers import (
    PROVIDER_DEMO,
    PROVIDER_LOCAL_ASSET,
)

# Providers that must never be packaged as scientific production.
NON_PROMOTABLE_PROVIDERS = frozenset({PROVIDER_DEMO, PROVIDER_LOCAL_ASSET})
NON_PROMOTABLE_MODEL_TYPES = frozenset({"demo", "heuristic"})

# Spatial output classes supported by the Stage-13 pipeline.
SPATIAL_VECTOR_POLYGONS = "VECTOR_POLYGONS"
SPATIAL_WELL_INTERVALS = "WELL_INTERVALS"
SPATIAL_CLASSIFIED_RASTER = "CLASSIFIED_RASTER"
SPATIAL_NONE = "NONE"

KNOWN_SPATIAL_TYPES = frozenset(
    {
        SPATIAL_VECTOR_POLYGONS,
        SPATIAL_WELL_INTERVALS,
        SPATIAL_CLASSIFIED_RASTER,
        SPATIAL_NONE,
        "",
    }
)


class ModelPackageError(ValueError):
    """Invalid model package manifest or registration request."""


@dataclass
class ModelPackageManifest:
    """Validated production (or test) model package metadata."""

    model_id: str
    model_version: str
    model_name: str
    capability: str
    provider: str
    artifact: str = ""
    checksum: str | None = None
    input_schema: dict[str, Any] = field(default_factory=dict)
    output_schema: dict[str, Any] = field(default_factory=dict)
    preprocessing_version: str = ""
    runtime: str = "python_callable"
    deterministic: bool = True
    spatial_output_type: str = SPATIAL_VECTOR_POLYGONS
    model_type: str = "ml"
    demo_only: bool = False
    scientific: bool = True
    provenance: dict[str, Any] = field(default_factory=dict)
    metadata: dict[str, Any] = field(default_factory=dict)

    def to_dict(self) -> dict[str, Any]:
        return {
            "model_id": self.model_id,
            "model_version": self.model_version,
            "model_name": self.model_name,
            "capability": self.capability,
            "provider": self.provider,
            "artifact": self.artifact,
            "checksum": self.checksum,
            "input_schema": dict(self.input_schema),
            "output_schema": dict(self.output_schema),
            "preprocessing_version": self.preprocessing_version,
            "runtime": self.runtime,
            "deterministic": self.deterministic,
            "spatial_output_type": self.spatial_output_type,
            "model_type": self.model_type,
            "demo_only": self.demo_only,
            "scientific": self.scientific,
            "provenance": dict(self.provenance),
            "metadata": dict(self.metadata),
        }


def _sha256_file(path: Path) -> str:
    h = hashlib.sha256()
    with path.open("rb") as handle:
        for chunk in iter(lambda: handle.read(1 << 20), b""):
            h.update(chunk)
    return h.hexdigest()


def load_manifest_dict(source: Path | str | dict[str, Any]) -> dict[str, Any]:
    """Load a package manifest from a dict or JSON path."""
    if isinstance(source, dict):
        return dict(source)
    path = Path(source)
    if not path.is_file():
        raise ModelPackageError(f"Manifest not found: {path}")
    try:
        data = json.loads(path.read_text(encoding="utf-8"))
    except (OSError, json.JSONDecodeError) as exc:
        raise ModelPackageError(f"Invalid manifest JSON: {exc}") from exc
    if not isinstance(data, dict):
        raise ModelPackageError("Manifest root must be a JSON object")
    return data


def _strict_bool(raw: dict[str, Any], key: str, default: bool) -> bool:
    """Parse a manifest boolean field strictly (H4-1).

    Python truthiness would turn ``"false"``/``"0"``/``"no"`` into True and
    silently invert the author's declaration. Only real booleans (or the
    unambiguous strings ``"true"``/``"false"``) are accepted.
    """
    value = raw.get(key, default)
    if isinstance(value, bool):
        return value
    if isinstance(value, str):
        s = value.strip().lower()
        if s in {"true", "false"}:
            return s == "true"
    raise ModelPackageError(f"{key} must be a boolean, got {value!r}")


def parse_model_package_manifest(
    source: Path | str | dict[str, Any],
    *,
    base_dir: Path | str | None = None,
) -> ModelPackageManifest:
    """Parse and validate a model package manifest (no catalog side effects)."""
    raw = load_manifest_dict(source)
    model_id = str(raw.get("model_id") or "").strip()
    model_version = str(raw.get("model_version") or raw.get("version") or "1").strip()
    model_name = str(raw.get("model_name") or raw.get("name") or model_id).strip()
    capability = str(raw.get("capability") or "").strip()
    provider = str(raw.get("provider") or "").strip()
    if not model_id:
        raise ModelPackageError("model_id is required")
    if not model_name:
        raise ModelPackageError("model_name is required")
    if not capability:
        raise ModelPackageError("capability is required")
    if not provider:
        raise ModelPackageError("provider is required")

    model_type = str(raw.get("model_type") or "ml").strip() or "ml"
    demo_only = _strict_bool(raw, "demo_only", False)
    scientific = _strict_bool(raw, "scientific", not demo_only)
    deterministic = _strict_bool(raw, "deterministic", True)
    spatial = str(raw.get("spatial_output_type") or SPATIAL_VECTOR_POLYGONS).strip()
    if spatial not in KNOWN_SPATIAL_TYPES:
        raise ModelPackageError(f"Unknown spatial_output_type: {spatial!r}")

    artifact = str(raw.get("artifact") or raw.get("artifact_uri") or "").strip()
    if artifact and base_dir is not None:
        art_path = Path(artifact)
        if not art_path.is_absolute():
            artifact = str((Path(base_dir) / art_path).resolve())

    checksum = raw.get("checksum")
    if checksum is not None:
        checksum = str(checksum).strip() or None

    input_schema = raw.get("input_schema") or {}
    output_schema = raw.get("output_schema") or {}
    if not isinstance(input_schema, dict) or not isinstance(output_schema, dict):
        raise ModelPackageError("input_schema and output_schema must be objects")
    metadata_raw = raw.get("metadata") or {}
    provenance_raw = raw.get("provenance") or {}
    if not isinstance(metadata_raw, dict) or not isinstance(provenance_raw, dict):
        raise ModelPackageError("metadata and provenance must be objects")

    # Enrich output_schema with spatial type when absent.
    if spatial and "spatial_output_type" not in output_schema:
        output_schema = {**output_schema, "spatial_output_type": spatial}

    return ModelPackageManifest(
        model_id=model_id,
        model_version=model_version,
        model_name=model_name,
        capability=capability,
        provider=provider,
        artifact=artifact,
        checksum=checksum,
        input_schema=dict(input_schema),
        output_schema=dict(output_schema),
        preprocessing_version=str(raw.get("preprocessing_version") or ""),
        runtime=str(raw.get("runtime") or "python_callable"),
        deterministic=deterministic,
        spatial_output_type=spatial or SPATIAL_NONE,
        model_type=model_type,
        demo_only=demo_only,
        scientific=scientific,
        provenance=dict(provenance_raw),
        metadata=dict(metadata_raw),
    )


def validate_model_package(
    manifest: ModelPackageManifest,
    *,
    require_artifact: bool = True,
    allow_non_scientific: bool = False,
) -> list[str]:
    """Return validation error messages (empty list means OK)."""
    errors: list[str] = []
    if manifest.provider in NON_PROMOTABLE_PROVIDERS and not allow_non_scientific:
        errors.append(
            f"provider {manifest.provider!r} cannot be registered as a production package"
        )
    if manifest.model_type in NON_PROMOTABLE_MODEL_TYPES and not allow_non_scientific:
        errors.append(
            f"model_type {manifest.model_type!r} cannot be a production package"
        )
    if manifest.demo_only and not allow_non_scientific:
        errors.append("demo_only packages cannot be registered as production")
    if manifest.scientific is False and not allow_non_scientific:
        # The author explicitly declared this package non-scientific; a later
        # promote must not be able to flip the label (H4-3b).
        errors.append("scientific=false packages cannot be registered as production")
    if require_artifact:
        if not manifest.artifact:
            errors.append("artifact path is required for production packages")
        else:
            path = Path(manifest.artifact)
            if not path.is_file():
                errors.append(f"artifact file missing: {path}")
            else:
                digest = _sha256_file(path)
                if manifest.checksum and manifest.checksum != digest:
                    errors.append(
                        f"checksum mismatch: manifest={manifest.checksum} file={digest}"
                    )
                elif not manifest.checksum:
                    # Fill for registration when valid.
                    manifest.checksum = digest
    if not manifest.input_schema and not allow_non_scientific:
        # Soft: production packages should declare inputs; warn as error for gate.
        errors.append("input_schema is required for production packages")
    return errors


def register_model_package(
    service,
    source: Path | str | dict[str, Any],
    *,
    base_dir: Path | str | None = None,
    status: str = "demo",
    require_artifact: bool = True,
    allow_non_scientific: bool = False,
) -> tuple[Model, ModelVersion]:
    """Validate a package and register Model + ModelVersion in the catalog.

    Default ``status="demo"`` so packages are not production until explicit
    :meth:`promote_model` (with safety gates). Pass ``status="production"``
    only for carefully validated packages (e.g. test fixtures that also pass
    promote gates).
    """
    if isinstance(source, (str, Path)) and base_dir is None:
        base_dir = Path(source).resolve().parent
    manifest = parse_model_package_manifest(source, base_dir=base_dir)
    errors = validate_model_package(
        manifest,
        require_artifact=require_artifact,
        allow_non_scientific=allow_non_scientific,
    )
    if errors:
        raise ModelPackageError("; ".join(errors))

    meta = {
        **manifest.metadata,
        "scientific": manifest.scientific,
        "spatial_output_type": manifest.spatial_output_type,
        "package": True,
    }
    model = service.register_model(
        model_id=manifest.model_id,
        model_name=manifest.model_name,
        model_type=manifest.model_type,
        capability=manifest.capability,
        provider=manifest.provider,
        status=status if status != "production" else "demo",
        metadata=meta,
        provenance=manifest.provenance,
        force_status=False,
    )
    # Re-register path for versions: get or create, with identity checks (H4-2).
    # Registering the same (model_id, model_version) with a DIFFERENT artifact
    # or schema must raise instead of silently returning the old version.
    try:
        existing_version = service.get_model_version(
            manifest.model_id, manifest.model_version
        )
    except Exception:
        existing_version = None
    if existing_version is not None:
        conflicts: list[str] = []
        if manifest.checksum and existing_version.checksum and manifest.checksum != existing_version.checksum:
            conflicts.append(f"checksum {existing_version.checksum} != {manifest.checksum}")
        if (
            manifest.input_schema
            and existing_version.input_schema
            and manifest.input_schema != existing_version.input_schema
        ):
            conflicts.append("input_schema differs")
        if (
            manifest.output_schema
            and existing_version.output_schema
            and manifest.output_schema != existing_version.output_schema
        ):
            conflicts.append("output_schema differs")
        if (
            manifest.preprocessing_version
            and existing_version.preprocessing_version
            and manifest.preprocessing_version != existing_version.preprocessing_version
        ):
            conflicts.append("preprocessing_version differs")
        if existing_version.deterministic != manifest.deterministic:
            conflicts.append("deterministic differs")
        if existing_version.demo_only != manifest.demo_only:
            conflicts.append("demo_only differs")
        if conflicts:
            raise ModelPackageError(
                f"identity conflict for {manifest.model_id}@{manifest.model_version}: "
                + "; ".join(conflicts)
            )
        version = existing_version
    else:
        version = service.register_model_version(
            manifest.model_id,
            model_version=manifest.model_version,
            artifact_uri=manifest.artifact,
            checksum=manifest.checksum,
            input_schema=manifest.input_schema,
            output_schema=manifest.output_schema,
            preprocessing_version=manifest.preprocessing_version,
            runtime=manifest.runtime,
            deterministic=manifest.deterministic,
            demo_only=manifest.demo_only,
            status="demo" if status == "production" else status,
            metadata=meta,
            provenance=manifest.provenance,
        )
    if status == "production":
        # Explicit promote with safety (will reject demo/heuristic).
        version = service.promote_model(manifest.model_id, manifest.model_version)
        model = service.get_model(manifest.model_id)
    return model, version


def can_promote_to_production(
    service, model_id: str, model_version: str, *, require_input_schema: bool = True
) -> tuple[bool, str]:
    """Return (ok, reason) for promote gates (shared with catalog.promote_model).

    ``require_input_schema`` is a PROMOTE-time requirement: a version promoted
    before the schema contract existed must not silently disappear from
    find_production_model on upgrade (Agent L P2) — reads keep the other
    gates but tolerate the legacy empty-schema case.
    """
    try:
        model = service.get_model(model_id)
        version = service.get_model_version(model_id, model_version)
    except Exception as exc:  # CatalogError
        return False, str(exc)
    if version.demo_only:
        return False, "demo_only model versions cannot be promoted to production"
    # Case/alias-insensitive checks: "Demo"/"DEMO"/"local-asset" must not
    # slip past exact-match allowlists (H4-3c).
    provider_norm = (model.provider or "").strip().casefold()
    if provider_norm in {p.strip().casefold() for p in NON_PROMOTABLE_PROVIDERS}:
        return False, f"provider {model.provider!r} is not promotable to production"
    type_norm = (model.model_type or "").strip().casefold()
    if type_norm in {t.strip().casefold() for t in NON_PROMOTABLE_MODEL_TYPES}:
        return False, f"model_type {model.model_type!r} is not promotable to production"
    if model.metadata.get("scientific") is False:
        return False, "model metadata marks scientific=False"
    if version.metadata.get("scientific") is False:
        return False, "version metadata marks scientific=False (H4-3b)"
    if require_input_schema and not version.input_schema:
        return False, "input_schema is required for production promotion (H5-b)"
    return True, "ok"

File: libs/prediction/src/model_package.cpp
#include "pwb/prediction/model_package.hpp"

#include <cstdint>
#include <filesystem>
#include <fstream>
#include <sstream>
#include <vector>

#include "pwb/domain/sha256.hpp"
#include "pwb/ingest/py_compat.hpp"
#include "python_compat.hpp"

namespace pwb::prediction {

namespace {

using detail::py_or;
using detail::py_repr;
using detail::py_str;
using detail::py_strip;
using detail::py_truthy;

const Json kNull;

const Json& get(const Json& obj, const char* key) {
    if (!obj.is_object()) return kNull;
    const auto it = obj.find(key);
    return it == obj.end() ? kNull : *it;
}

std::string ascii_lower(std::string s) {
    for (char& c : s) {
        if (c >= 'A' && c <= 'Z') c = static_cast<char>(c + 32);
    }
    return s;
}

// _strict_bool(raw, key, default): bool passthrough or "true"/"false"
// (strip + lower); anything else -> ModelPackageError with repr().
bool strict_bool(const Json& raw, const char* key, bool dflt) {
    const Json& v = raw.contains(key) ? raw[key] : Json(dflt);
    if (v.is_boolean()) return v.get<bool>();
    if (v.is_string()) {
        const std::string s =
            ascii_lower(py_strip(v.get_ref<const std::string&>()));
        if (s == "true") return true;
        if (s == "false") return false;
    }
    throw ModelPackageError(std::string(key) + " must be a boolean, got " +
                            py_repr(v));
}

// Python 'utf-8' strict decode error report: position + reason class.
// Only the structural classes Python emits for well-formed-enough streams
// are reproduced (invalid start byte / invalid continuation byte /
// unexpected end of data) — the oracle freezes the concrete case.
struct Utf8Error {
    size_t pos;
    size_t end;       // exclusive; > pos+1 means Python's "N-M" range form
    unsigned byte;    // offending byte at pos (invalid start/continuation)
    const char* reason;
};

std::optional<Utf8Error> utf8_error_at(std::string_view bytes) {
    for (size_t i = 0; i < bytes.size();) {
        const auto c = static_cast<unsigned char>(bytes[i]);
        if (c < 0x80) { ++i; continue; }
        size_t len = 0;
        if (c >= 0xC2 && c <= 0xDF) len = 2;
        else if (c >= 0xE0 && c <= 0xEF) len = 3;
        else if (c >= 0xF0 && c <= 0xF4) len = 4;
        else return Utf8Error{i, i + 1, c, "invalid start byte"};
        if (i + len > bytes.size()) {
            return Utf8Error{i, bytes.size(), c, "unexpected end of data"};
        }
        for (size_t k = 1; k < len; ++k) {
            if ((static_cast<unsigned char>(bytes[i + k]) & 0xC0) != 0x80) {
                return Utf8Error{i + k, i + k + 1,
                                 static_cast<unsigned char>(bytes[i + k]),
                                 "invalid continuation byte"};
            }
        }
        i += len;
    }
    return std::nullopt;
}

std::string unicode_decode_message(const Utf8Error& e) {
    char buf[160];
    if (e.end > e.pos + 1) {
        std::snprintf(buf, sizeof buf,
                      "'utf-8' codec can't decode bytes in position %zu-%zu: %s",
                      e.pos, e.end - 1, e.reason);
    } else {
        std::snprintf(buf, sizeof buf,
                      "'utf-8' codec can't decode byte 0x%02x in position %zu: %s",
                      e.byte, e.pos, e.reason);
    }
    return std::string(buf);
}

}  // namespace

Json ModelPackageManifest::to_dict() const {
    return Json{
        {"model_id", model_id},
        {"model_version", model_version},
        {"model_name", model_name},
        {"capability", capability},
        {"provider", provider},
        {"artifact", artifact},
        {"checksum", checksum ? Json(*checksum) : Json()},
        {"input_schema", input_schema},
        {"output_schema", output_schema},
        {"preprocessing_version", preprocessing_version},
        {"runtime", runtime},
        {"deterministic", deterministic},
        {"spatial_output_type", spatial_output_type},
        {"model_type", model_type},
        {"demo_only", demo_only},
        {"scientific", scientific},
        {"provenance", provenance},
        {"metadata", metadata},
    };
}

ModelPackageManifest ModelPackageManifest::from_dict(const Json& d) {
    ModelPackageManifest m;
    auto str_or = [&](const char* k, std::string& dst) {
        if (d.contains(k) && d[k].is_string()) dst = d[k].get<std::string>();
    };
    auto bool_or = [&](const char* k, bool& dst) {
        if (d.contains(k) && d[k].is_boolean()) dst = d[k].get<bool>();
    };
    auto obj_or = [&](const char* k, Json& dst) {
        if (d.contains(k) && d[k].is_object()) dst = d[k];
    };
    str_or("model_id", m.model_id);
    str_or("model_version", m.model_version);
    str_or("model_name", m.model_name);
    str_or("capability", m.capability);
    str_or("provider", m.provider);
    str_or("artifact", m.artifact);
    if (d.contains("checksum") && d["checksum"].is_string()) {
        m.checksum = d["checksum"].get<std::string>();
    }
    obj_or("input_schema", m.input_schema);
    obj_or("output_schema", m.output_schema);
    str_or("preprocessing_version", m.preprocessing_version);
    str_or("runtime", m.runtime);
    bool_or("deterministic", m.deterministic);
    str_or("spatial_output_type", m.spatial_output_type);
    str_or("model_type", m.model_type);
    bool_or("demo_only", m.demo_only);
    bool_or("scientific", m.scientific);
    obj_or("provenance", m.provenance);
    obj_or("metadata", m.metadata);
    return m;
}

Json load_manifest_dict(const Json& source) {
    if (source.is_object()) return source;
    if (!source.is_string()) {
        throw ModelPackageError(
            "expected str, bytes or os.PathLike object, not '" +
            detail::py_type_name(source) + "'");
    }
    const std::string path_str = source.get_ref<const std::string&>();
    const std::filesystem::path path(path_str);
    if (!std::filesystem::is_regular_file(path)) {
        throw ModelPackageError("Manifest not found: " + path_str);
    }
    std::ifstream in(path, std::ios::binary);
    std::ostringstream buf;
    buf << in.rdbuf();
    const std::string bytes = buf.str();
    // Path.read_text(encoding="utf-8") — strict; the decode failure escapes
    // unwrapped as UnicodeDecodeError (not ModelPackageError).
    if (const auto err = utf8_error_at(bytes)) {
        throw UnicodeDecodeError(unicode_decode_message(*err));
    }
    Json data;
    try {
        data = Json::parse(bytes);
    } catch (const std::exception& exc) {
        // Detail text differs from CPython's json error wording — the
        // "Invalid manifest JSON: " prefix is the frozen contract (D9).
        throw ModelPackageError(std::string("Invalid manifest JSON: ") +
                                exc.what());
    }
    if (!data.is_object()) {
        throw ModelPackageError("Manifest root must be a JSON object");
    }
    return data;
}

ModelPackageManifest parse_model_package_manifest(const Json& source,
                                                  const Json& base_dir) {
    const Json raw = load_manifest_dict(source);
    const std::string model_id =
        py_strip(py_str(py_or(get(raw, "model_id"), Json(""))));
    const std::string model_version = py_strip(py_str(
        py_or(get(raw, "model_version"), py_or(get(raw, "version"), Json("1")))));
    const std::string model_name = py_strip(py_str(py_or(
        py_or(get(raw, "model_name"), get(raw, "name")), Json(model_id))));
    const std::string capability =
        py_strip(py_str(py_or(get(raw, "capability"), Json(""))));
    const std::string provider =
        py_strip(py_str(py_or(get(raw, "provider"), Json(""))));
    if (model_id.empty()) throw ModelPackageError("model_id is required");
    if (model_name.empty()) throw ModelPackageError("model_name is required");
    if (capability.empty()) throw ModelPackageError("capability is required");
    if (provider.empty()) throw ModelPackageError("provider is required");

    std::string model_type =
        py_strip(py_str(py_or(get(raw, "model_type"), Json("ml"))));
    if (model_type.empty()) model_type = "ml";
    const bool demo_only = strict_bool(raw, "demo_only", false);
    const bool scientific = strict_bool(raw, "scientific", !demo_only);
    const bool deterministic = strict_bool(raw, "deterministic", true);
    const std::string spatial = py_strip(
        py_str(py_or(get(raw, "spatial_output_type"),
                     Json(std::string(kSpatialVectorPolygons)))));
    if (!kKnownSpatialTypes.count(spatial)) {
        throw ModelPackageError("Unknown spatial_output_type: " +
                                py_repr(Json(spatial)));
    }

    std::string artifact = py_strip(py_str(
        py_or(get(raw, "artifact"), py_or(get(raw, "artifact_uri"), Json("")))));
    if (!artifact.empty() && !base_dir.is_null()) {
        const std::filesystem::path art_path(artifact);
        if (!art_path.is_absolute()) {
            const std::filesystem::path base(
                base_dir.is_string() ? base_dir.get_ref<const std::string&>()
                                     : py_str(base_dir));
            // Path.resolve(strict=False) ~= weakly_canonical (D9).
            artifact = std::filesystem::weakly_canonical(base / art_path)
                           .generic_string();
        }
    }

    std::optional<std::string> checksum;
    const Json& checksum_v = get(raw, "checksum");
    if (!checksum_v.is_null()) {
        const std::string s = py_strip(py_str(checksum_v));
        if (!s.empty()) checksum = s;
    }

    const Json input_schema =
        py_or(get(raw, "input_schema"), Json::object());
    const Json output_schema_raw =
        py_or(get(raw, "output_schema"), Json::object());
    if (!input_schema.is_object() || !output_schema_raw.is_object()) {
        throw ModelPackageError(
            "input_schema and output_schema must be objects");
    }
    const Json metadata =
        py_or(get(raw, "metadata"), Json::object());
    const Json provenance =
        py_or(get(raw, "provenance"), Json::object());
    if (!metadata.is_object() || !provenance.is_object()) {
        throw ModelPackageError("metadata and provenance must be objects");
    }

    Json output_schema = output_schema_raw;
    if (!spatial.empty() && !output_schema.contains("spatial_output_type")) {
        output_schema["spatial_output_type"] = spatial;
    }

    ModelPackageManifest m;
    m.model_id = model_id;
    m.model_version = model_version;
    m.model_name = model_name;
    m.capability = capability;
    m.provider = provider;
    m.artifact = artifact;
    m.checksum = checksum;
    m.input_schema = input_schema;
    m.output_schema = output_schema;
    m.preprocessing_version =
        py_str(py_or(get(raw, "preprocessing_version"), Json("")));
    m.runtime = py_str(py_or(get(raw, "runtime"), Json("python_callable")));
    m.deterministic = deterministic;
    m.spatial_output_type = spatial.empty() ? std::string(kSpatialNone) : spatial;
    m.model_type = model_type;
    m.demo_only = demo_only;
    m.scientific = scientific;
    m.provenance = provenance;
    m.metadata = metadata;
    return m;
}

Json validate_model_package(ModelPackageManifest& manifest,
                            bool require_artifact,
                            bool allow_non_scientific) {
    Json errors = Json::array();
    if (!allow_non_scientific &&
        kNonPromotableProviders.count(manifest.provider)) {
        errors.push_back("provider " + py_repr(Json(manifest.provider)) +
                         " cannot be registered as a production package");
    }
    if (!allow_non_scientific &&
        kNonPromotableModelTypes.count(manifest.model_type)) {
        errors.push_back("model_type " + py_repr(Json(manifest.model_type)) +
                         " cannot be a production package");
    }
    if (!allow_non_scientific && manifest.demo_only) {
        errors.push_back(
            "demo_only packages cannot be registered as production");
    }
    if (!allow_non_scientific && !manifest.scientific) {
        errors.push_back(
            "scientific=false packages cannot be registered as production");
    }
    if (require_artifact) {
        if (manifest.artifact.empty()) {
            errors.push_back(
                "artifact path is required for production packages");
        } else {
            const std::filesystem::path path(manifest.artifact);
            if (!std::filesystem::is_regular_file(path)) {
                errors.push_back("artifact file missing: " + manifest.artifact);
            } else {
                const auto digest = pwb::domain::Sha256::of_file(path);
                if (!digest) {
                    throw std::runtime_error("artifact vanished: " +
                                             manifest.artifact);
                }
                if (manifest.checksum && *manifest.checksum != *digest) {
                    errors.push_back("checksum mismatch: manifest=" +
                                     *manifest.checksum + " file=" + *digest);
                } else if (!manifest.checksum) {
                    manifest.checksum = *digest;
                }
            }
        }
    }
    if (!allow_non_scientific && !py_truthy(manifest.input_schema)) {
        errors.push_back("input_schema is required for production packages");
    }
    return errors;
}

}  // namespace pwb::prediction
